File: src/main/java/nz/co/searchwellington/controllers/admin/AdminUrlBuilder.java
package nz.co.searchwellington.controllers.admin;

import nz.co.searchwellington.model.Feed;
import nz.co.searchwellington.model.FrontendFeedNewsitem;
import nz.co.searchwellington.model.SiteInformation;
import nz.co.searchwellington.model.frontend.FrontendFeed;
import nz.co.searchwellington.model.frontend.FrontendResource;
import nz.co.searchwellington.model.frontend.FrontendWebsite;
import nz.co.searchwellington.urls.UrlBuilder;
import nz.co.searchwellington.urls.UrlParameterEncoder;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import com.google.common.base.Strings;

@Component
public class AdminUrlBuilder {
	
	private SiteInformation siteInformation;
	private UrlBuilder urlBuilder;
	
	@Autowired
	public AdminUrlBuilder(SiteInformation siteInformation, UrlBuilder urlBuilder) {		
		this.siteInformation = siteInformation;
		this.urlBuilder = urlBuilder;
	}
	
	public String getResourceEditUrl(FrontendResource resource) {
		if (resource.getId() > 0) {
			return siteInformation.getUrl() + "/edit?resource=" + resource.getId();
		} else if (!Strings.isNullOrEmpty(resource.getUrlWords())) {
			return siteInformation.getUrl() + "/edit?resource=" + UrlParameterEncoder.encode(resource.getUrlWords());
		}
		return null;
	}
	
	public String getResourceDeleteUrl(FrontendResource resource) {
		return siteInformation.getUrl() + "/delete?resource=" + resource.getId();		
	}
	
	public String getResourceCheckUrl(FrontendResource resource) {
		return siteInformation.getUrl() + "/admin/linkchecker/add?resource=" + resource.getId();
	}
	
	public String getViewSnapshotUrl(FrontendResource resource) {
		final String resourceUrl = siteInformation.getUrl() + "/" + resource.getUrlWords();
		return resourceUrl + "/viewsnapshot";	
	}
	
	public String getFeednewsItemAcceptUrl(FrontendFeed feed, FrontendFeedNewsitem feednewsitem) {
		return siteInformation.getUrl() + "/edit/accept?feed=" + feed.getUrlWords() + "&url=" + UrlParameterEncoder.encode(feednewsitem.getUrl());		
	}
	
	public String getAcceptAllFromFeed(Feed feed) {
		return siteInformation.getUrl() + "/admin/feed/acceptall?feed=" + feed.getUrlWords();		
	}
	
	public String getDecacheFeed(Feed feed) {
		return siteInformation.getUrl() + "/admin/feed/decache?feed=" + feed.getUrlWords();		
	}
	
	public String getFeedNewsitemSuppressUrl(FrontendFeedNewsitem feednewsitem) {
		return siteInformation.getUrl() + "/supress/supress?url=" + UrlParameterEncoder.encode(feednewsitem.getUrl());
	}
	
	public String getFeedNewsitemUnsuppressUrl(FrontendFeedNewsitem feednewsitem) {
		return siteInformation.getUrl() + "/supress/unsupress?url=" + UrlParameterEncoder.encode(feednewsitem.getUrl());
	}
	
	public String getPublisherAutoGatherUrl(FrontendWebsite resource) {
		final String resourceUrl = urlBuilder.getResourceUrl(resource);
		if (resourceUrl != null) {
			return resourceUrl + "/gather";
		}
		return null;
	}
	
	public String getAddTagUrl() {
		return siteInformation.getUrl() + "/edit/tag/submit";
	}
	
}

File: src/main/scala/nz/co/searchwellington/controllers/admin/AdminUrlBuilder.scala
package nz.co.searchwellington.controllers.admin

import nz.co.searchwellington.feeds.whakaoko.model.Subscription
import nz.co.searchwellington.model.frontend.{FrontendFeed, FrontendNewsitem, FrontendResource, FrontendWatchlist, FrontendWebsite}
import nz.co.searchwellington.model._
import nz.co.searchwellington.urls.{UrlBuilder, UrlParameterEncoder}
import org.springframework.beans.factory.annotation.{Autowired, Value}
import org.springframework.stereotype.Component

@Component
class AdminUrlBuilder @Autowired()(urlBuilder: UrlBuilder, @Value("${whakaoko.url}") whakaokoUrl: String) {

  def adminPage(): String = "/admin"

  def getResourceEditUrl(resource: FrontendResource): String = {
    resource match {
      case f: FrontendFeed =>
        "/edit-feed/" + f.id
      case n: FrontendNewsitem =>
        "/edit-newsitem/" + n.id
      case w: FrontendWebsite =>
        "/edit-website/" + w.id
      case l: FrontendWatchlist =>
        "/edit-watchlist/" + l.id
    }
  }
  def getResourceEditUrl(resource: Resource): String = {
    resource match {
      case f: Feed =>
        "/edit-feed/" + f.id
      case n: Newsitem =>
        "/edit-newsitem/" + n.id
      case w: Website =>
        "/edit-website/" + w.id
      case l: Watchlist =>
        "/edit-watchlist/" + l.id
    }
  }

  def getResourceDeleteUrl(resource: FrontendResource): String = {
    "/delete-resource/" + resource.getId
  }

  def getSaveUrl: String = {
    "/save"
  }

  def getResourceCheckUrl(resource: FrontendResource): String = {
    "/check-resource/" + resource.getId
  }

  def getViewSnapshotUrl(resource: FrontendResource): String = {
    "/" + resource.getUrlWords + "/viewsnapshot"
  }

  def getFeednewsItemAcceptUrl(feed: FrontendFeed, newsitem: FrontendNewsitem): String = {
    "/accept-feed-item?feed=" + feed.urlWords + "&url=" + UrlParameterEncoder.encode(newsitem.getUrl)
  }

  def getAcceptAllFromFeed(feed: FrontendFeed): String = {
    urlBuilder.getFeedUrl(feed) + "/accept-all"
  }

  def getFeedNewsitemSuppressUrl(newsitem: FrontendNewsitem): String = {
    "/suppress/suppress?url=" + UrlParameterEncoder.encode(newsitem.getUrl)
  }

  def getFeedNewsitemUnsuppressUrl(newsitem: FrontendNewsitem): String = {
    "/suppress/unsuppress?url=" + UrlParameterEncoder.encode(newsitem.getUrl)
  }

  def getPublisherAutoGatherUrl(publisher: FrontendWebsite): String = {
    "/admin/gather/" + publisher.id
  }

  def getAddTagUrl: String = "/new-tag"

  def getEditTagUrl(tag: Tag): String = {
    "/edit-tag/" + tag.id
  }

  def deleteTagUrl(tag: Tag): String = "/delete-tag/" + tag.id

  def getWhakaokoPreviewUrl(subscription: Subscription): String = {
    if (subscription != null) {
      whakaokoUrl + "/ui/" + "/subscriptions/" + subscription.id
    } else {
      ""
    }
  }

}
